### ISSUE TITLE
Datastore upsert update not working with index _id

### ISSUE THREAD

**Issue by digital-codes** [ckan/ckan#6793]
**CKAN version** 2.9.5

**Describe the bug**
datastore/upsert method=update does not work with _id 

**Steps to reproduce**
Create resource in datastore
insert values
read values => resulting records have unique field _id, as expected
update command with _id fails on "undefined primary_key"
{
    "error": {
        "__type": "Validation Error",
        "table": [
            "table does not have a unique key defined"
        ]
    },

**Expected behavior**
Record number _id can be used as primary key, as in regular SQL database

**Additional details**
If possible, please provide the full stack trace of the error raised,  or add screenshots to help explain your problem.

Use case: update resource with timeseries data which by design do not come with an inherent primary key. So primary key must be database id. However, this cannot be enforced on datastore_create()

Maybe you consider this not a bug - but we have an id in the data, so not allowing to use it is quite close ...

**Comment by TomeCirun** [comment]
Hey, @digital-codes,  I've tested this and it doesn't work as you think. 

When using `datastore_create` action, you have `param: primary_key` this `primary_key` is optional. 

When you set your `fields/columns` e.g like this: `"fields": [{"id": "name"}, {"id": "age"}, {"id": "email"} ]` and your records like: `"records": [ {"name": "Andreas", "age": 30, "email": "andreas@digital-codes.com"} ]` you can set `primary_key` to be your `"email"` this would be your `unique_key` in your `fields/columns`.

Then you can use `datastore_upsert` (lets say you want to update your name) like this: 

curl -X POST http://localhost:5000/api/3/action/datastore_upsert -H "Authorization: your-key" -d '{"resource_id": "{resource.id}", "force": "True", "records": [{"name": "Kugel", "email": "andreas@digital-codes.com"}], "method": "update" }'.

If you don’t set your `primary_key` when creating, you cannot use `datastore_upsert` as all three methods require a `unique key`.

Thanks

**Comment by wardi** [comment]
I agree with the reporter. The `_id` field should be usable as a primary key for `datastore_upsert`. I'm happy to review a PR or help work on a fix for this issue.

**Comment by digital-codes** [comment]
Thanks for taking this issue. I'm not too familiar with CKAN internals nor Postgres, but if I'll try to help. I looked at backend/postgres.py and there is 

`elif method in [_UPDATE, _UPSERT]:
        unique_keys = _get_unique_key(context, data_dict)
        if len(unique_keys) < 1:
            raise ValidationError({
                'table': [u'table does not have a unique key defined']
            })
` 
which is probably giving the error. But I don't know why postgres doesn't accept the _id as a unique key. 

I looked further and *create_table* goes like:

`
def create_table(context: Context, data_dict: dict[str, Any]):
    '''Creates table, columns and column info (stored as comments).
    :param resource_id: The resource ID (i.e. postgres table name)
    :type resource_id: string
    :param fields: details of each field/column, each with properties:
        id - field/column name
        type - optional, otherwise it is guessed from the first record
        info - some field/column properties, saved as a JSON string in postgres
            as a column comment. e.g. "type_override", "label", "notes"
    :type fields: list of dicts
    :param records: records, of which the first is used when a field type needs
        guessing.
    :type records: list of dicts
    '''

    datastore_fields = [
        {'id': '_id', 'type': 'serial primary key'},
        {'id': '_full_text', 'type': 'tsvector'},
    ]

` 
Which seems to set _id as a primary key. Or is this a type and it should be _primary_key_ instead of _primary_  and _key_ ?

**Comment by digital-codes** [comment]
Hmmm, maybe it's the way how the unique keys are looked up:

```
def _get_unique_key(context: Context, data_dict: dict[str, Any]) -> list[str]:
    sql_get_unique_key = '''
    SELECT
        a.attname AS column_names
    FROM
        pg_class t,
        pg_index idx,
        pg_attribute a
    WHERE
        t.oid = idx.indrelid
        AND a.attrelid = t.oid
        AND a.attnum = ANY(idx.indkey)
        AND t.relkind = 'r'
        AND idx.indisunique = true
        AND idx.indisprimary = false
        AND t.relname = %s
    '''

```
The line 
        AND idx.indisprimary = false
will probably exclude the _id ...

**Comment by TomeCirun** [comment]
> The line AND idx.indisprimary = false will probably exclude the _id ...

Yep thats true

**Comment by TomeCirun** [comment]
Hey @wardi, I'd like to work on this issue, currently, I have a fix though might not seem the best solution, if you don't agree with this please give me some pointers.

Here is what I've done: 
I removed this line: `AND idx.indisprimary = false` from `_get_unique_key()` 
with that line removed, from now we will always receive the `_id` when calling `_get_unique_key()` 

i will comment my changes with `# this line is added by me` so you will know what has been added
```python
for num, record in enumerate(records):
      # remove the _id if not provided when calling update method
      if '_id' in unique_keys and '_id' not in record: # this line is added by me
          unique_keys.remove('_id') # this line is added by me
      # all key columns have to be defined
      missing_fields = [field for field in unique_keys
                        if field not in record]
      if missing_fields:
          raise ValidationError({
              'key': [u'''fields "{fields}" are missing
                  but needed as key'''.format(
                      fields=', '.join(missing_fields))]
          })

      for field in fields:
          value = record.get(field['id'])
          if value is not None and field['type'].lower() == 'nested':
              # a tuple with an empty second value
              record[field['id']] = (json.dumps(value), '')

      if "_id" in record: # this line is added by me
          # _id is not a field_name so we need to append
          field_names.append('_id') # this line is added by me
      non_existing_filed_names = [field for field in record
                                  if field not in field_names]
````

the original code you can find here https://github.com/ckan/ckan/blob/2e63e7fbaf7208f122e8c90a419f7c71485423b9/ckanext/datastore/backend/postgres.py#L1170

Thanks

**Comment by wardi** [comment]
@TomeCirun thanks, would you mind creating a pull request instead?
